pandas/core/indexers.py

```diff
@@ -307,3 +312,62 @@ def check_bool_array_indexer(array: AnyArrayLike, mask: AnyArrayLike) -> np.ndar
     if len(result) != len(array):
         raise IndexError(f"Item wrong length {len(result)} instead of {len(array)}.")
     return result
+
+
+def check_array_indexer(array, indexer) -> np.ndarray:
+    """
+    Check if `indexer` is a valid array indexer for `array`.
+
+    `array` and `indexer` are checked to have the same length, and the
+    dtype is validated. If it is an integer or boolean ExtensionArray, it is
+    checked if there are missing values present, and it is converted to
+    the appropriate numpy array.
+
+    .. versionadded:: 1.0.0
+
+    Parameters
+    ----------
+    array : array
+        The array that's being indexed (only used for the length).
+    indexer : array-like
+        The array-like that's used to index.
+
+    Returns
+    -------
+    numpy.ndarray
+        The validated indexer.
+
+    Raises
+    ------
+    IndexError
+        When the lengths don't match.
+    ValueError
+        When `indexer` cannot be converted to a numpy ndarray.
+
+    """
```

Comment:
does this subsume check_bool_array_indexer? if not why not?

Reply:
> does this subsume check_bool_array_indexer? if not why not?

Yes, as mentioned elsewhere as well, if we agree on the general idea of this PR, need to clean this up. I don't think there is a reason to keep both, so was planning to replace `check_bool_array_indexer` completely with check_bool_array_indexer

Reply:
k, agree on the general idea, if you can merge them would be ideal.